db/db_impl_compaction_flush.cc

```diff
@@ -777,6 +777,26 @@ Status DBImpl::CompactRange(const CompactRangeOptions& options,
   return s;
 }
 
+  class SnapshotListFetchCallbackImpl : public SnapshotListFetchCallback {
+    public:
+    SnapshotListFetchCallbackImpl(InstrumentedMutex* mutex, SnapshotList* snapshotlist, Logger* info_log) :mutex_(mutex), snapshotlist_(snapshotlist), info_log_(info_log) {}
+    virtual void Refresh(std::vector<SequenceNumber>* snapshots,
+                         SequenceNumber max) override {
+      mutex_->Lock();
+      size_t prev = snapshots->size();
+      *snapshots = std::move(snapshotlist_->GetAll(nullptr, max));
+      size_t now = snapshots->size();
+      ROCKS_LOG_DEBUG(info_log_,
+                      "Compaction snapshot count refreshed from %zu to %zu",
+                      prev, now);
+      mutex_->Unlock();
+    }
+    private:
+    InstrumentedMutex* mutex_;
+    SnapshotList* snapshotlist_;
```

Comment:
I feel a little bit nervous taking two internal DBImpl object, DB mutex and snapshot list here. Do we consider to keep them be encapsulated in DBImpl, and call a function of DBImpl to update the list?

Reply:
All such function would do is this:
```
        mutex_->Lock();
        *snapshots = std::move(snapshotlist_->GetAll(nullptr, max));
        mutex_->Unlock();
```
Let me know if you prefer adding a new function for that.

Reply:
Sounds like a quite typical function in DBImpl. I prefer adding it.